provider/ansible.py

```diff
@@ -167,11 +167,18 @@ def distro_install(packages="ansible"):
     policy_map = {"distro_install": distro_install,
                   "python_install": python_install}
 
+    ansible_install_policy = params['ansible_install_policy']
     package_list = params.get_list("package_list", 'sshpass')
     try:
         path.find_command('ansible-playbook')
-    except path.CmdNotFoundError:
-        ansible_install_policy = params['ansible_install_policy']
+        if ansible_install_policy == 'python_install':
+            check_cmd = r"pip freeze | grep -v ansible-core | grep -q ansible="
+        elif params.get("ansible_check_cmd"):
+            check_cmd = params.get("ansible_check_cmd")
+        if "check_cmd" in locals():
+            LOG_JOB.debug(f"Is full ansible version installed: '{check_cmd}'")
+            process.run(check_cmd, verbose=False, shell=True)
+    except (path.CmdNotFoundError, process.CmdError):
```

Comment:
```
    try:      
        if ansible_install_policy == 'python_install':
            check_cmd = r"pip freeze | grep -v ansible-core | grep -q ansible="
        elif ansible_install_policy == 'distro_install':
            check_cmd = params.get("ansible_check_cmd")
        
        if check_cmd:
            LOG_JOB.debug(f"Is full ansible version installed: '{check_cmd}'")
            process.run(check_cmd, verbose=False, shell=True)
        else:
            path.find_command('ansible-playbook')
    except (path.CmdNotFoundError, process.CmdError):
        ...
        ...
````

Reply:
@yfu I slightly updated it to handle situation if `check_cmd` variable is not defined at all (meaning that none or not valid install policy was passed) and not raise misleading exception about undefined variable :slightly_smiling_face: 

```
    try:
        if ansible_install_policy == 'python_install':
            check_cmd = r"pip freeze | grep -v ansible-core | grep -q ansible="
        elif ansible_install_policy == 'distro_install':
            check_cmd = params.get("ansible_check_cmd")
        else:
            LOG_JOB.error(f"No valid install policy: {ansible_install_policy}.")
            return False
        if check_cmd:
            LOG_JOB.debug(f"Is full ansible version installed: '{check_cmd}'")
            process.run(check_cmd, verbose=False, shell=True)
        else:
            path.find_command('ansible-playbook')
``` 

Let me test it to see if everything works as expected :wink:

Reply:
@yanan-fu @PaulYuuu Test including the latest changes look good :slightly_smiling_face: 

a) Test with `python_install` policy and `ansible-core` previously installed. Installation of Ansible is: **OK**

```
[stdlog] 2023-05-25 09:01:16,363 avocado.utils.process INFO | Running 'which pip'
[stdlog] 2023-05-25 09:01:16,400 avocado.utils.process DEBUG| [stdout] /usr/local/bin/pip
[stdlog] 2023-05-25 09:01:16,401 avocado.utils.process INFO | Command 'which pip' finished with 0 after 0.009276690s
[stdlog] 2023-05-25 09:01:16,402 avocado.test DEBUG| Is full ansible version installed: 'pip freeze | grep -v ansible-core | grep -q ansible='
[stdlog] 2023-05-25 09:01:19,756 avocado.utils.process INFO | Running 'which pip'
[stdlog] 2023-05-25 09:01:19,774 avocado.utils.process DEBUG| [stdout] /usr/local/bin/pip
[stdlog] 2023-05-25 09:01:19,775 avocado.utils.process INFO | Command 'which pip' finished with 0 after 0.002594150s
[stdlog] 2023-05-25 09:01:19,776 avocado.utils.process INFO | Running 'pip install ansible'
[stdlog] 2023-05-25 09:01:22,363 avocado.utils.process DEBUG| [stdout] Collecting ansible
[stdlog] 2023-05-25 09:01:22,899 avocado.utils.process DEBUG| [stdout]   Using cached ansible-7.6.0-py3-none-any.whl (43.8 MB)
(...)
[stdlog] 2023-05-25 09:01:24,073 avocado.utils.process DEBUG| [stdout] Installing collected packages: ansible
[stdlog] 2023-05-25 09:01:55,754 avocado.utils.process DEBUG| [stdout] Successfully installed ansible-7.6.0
[stdlog] 2023-05-25 09:01:55,755 avocado.utils.process DEBUG| [stderr] WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
[stdlog] 2023-05-25 09:01:55,889 avocado.utils.process INFO | Command 'pip install ansible' finished with 0 after 36.095880663s
```

b) Test with `python_install` policy and no `ansible` version installed. Installation of Ansible is: **OK**

```
[stdlog] 2023-05-25 09:07:40,178 avocado.utils.process INFO | Running 'which pip'
[stdlog] 2023-05-25 09:07:40,208 avocado.utils.process DEBUG| [stdout] /usr/local/bin/pip
[stdlog] 2023-05-25 09:07:40,209 avocado.utils.process INFO | Command 'which pip' finished with 0 after 0.009336787s
[stdlog] 2023-05-25 09:07:40,210 avocado.test DEBUG| Is full ansible version installed: 'pip freeze | grep -v ansible-core | grep -q ansible='
[stdlog] 2023-05-25 09:07:41,989 avocado.virttest.ip_sniffing WARNI| IPv6 address sniffing is not supported yet by using TShark, please fallback to use other sniffers by uninstalling TShark when testing with IPv6
[stdlog] 2023-05-25 09:07:43,472 avocado.utils.process INFO | Running 'which pip'
[stdlog] 2023-05-25 09:07:43,490 avocado.utils.process DEBUG| [stdout] /usr/local/bin/pip
[stdlog] 2023-05-25 09:07:43,491 avocado.utils.process INFO | Command 'which pip' finished with 0 after 0.002893661s
[stdlog] 2023-05-25 09:07:43,492 avocado.utils.process INFO | Running 'pip install ansible'
[stdlog] 2023-05-25 09:07:45,585 avocado.utils.process DEBUG| [stdout] Collecting ansible
[stdlog] 2023-05-25 09:07:46,094 avocado.utils.process DEBUG| [stdout]   Using cached ansible-7.6.0-py3-none-any.whl (43.8 MB)
[stdlog] 2023-05-25 09:07:47,017 avocado.utils.process DEBUG| [stdout] Collecting ansible-core~=2.14.6 (from ansible)
[stdlog] 2023-05-25 09:07:47,071 avocado.utils.process DEBUG| [stdout]   Using cached ansible_core-2.14.6-py3-none-any.whl (2.2 MB)
(...)
[stdlog] 2023-05-25 09:07:47,738 avocado.utils.process DEBUG| [stdout] Installing collected packages: ansible-core, ansible
[stdlog] 2023-05-25 09:08:21,713 avocado.utils.process DEBUG| [stdout] Successfully installed ansible-7.6.0 ansible-core-2.14.6
```

c) Test with `distro_install` policy and no `ansible` version installed. No `ansible_check_cmd` is defined. 

Installation of Ansible is: **OK**

```
[stdlog] 2023-05-25 09:14:55,743 avocado.utils.process INFO | Running 'rpm -q ansible'
[stdlog] 2023-05-25 09:14:55,799 avocado.utils.process DEBUG| [stdout] package ansible is not installed
[stdlog] 2023-05-25 09:14:55,801 avocado.utils.process INFO | Command 'rpm -q ansible' finished with 1 after 0.037041742s
[stdlog] 2023-05-25 09:14:55,802 avocado.utils.process INFO | Running '/usr/bin/dnf -y install ansible'
[stdlog] 2023-05-25 09:14:58,045 avocado.utils.process DEBUG| [stdout] Updating Subscription Management repositories.
[stdlog] 2023-05-25 09:14:58,047 avocado.utils.process DEBUG| [stdout] Unable to read consumer identity
(...)
[stdlog] 2023-05-25 09:15:27,508 avocado.utils.process DEBUG| [stdout] Complete!
[stdlog] 2023-05-25 09:15:27,808 avocado.utils.process INFO | Command '/usr/bin/dnf -y install ansible' finished with 0 after 31.991125356s

```

d) Test with `distro_install` policy and `ansible-core` installed. `ansible_check_cmd` set to `rpm -q ansible`. 

Installation of Ansible is: **OK**

```
[stdlog] 2023-05-25 09:22:06,874 avocado.test DEBUG| Is full ansible version installed: 'rpm -q ansible'
[stdlog] 2023-05-25 09:22:10,183 avocado.utils.software_manager DEBUG| dnf version: 4.14.0
[stdlog] 2023-05-25 09:22:10,184 avocado.utils.process INFO | Running 'rpm -q ansible'
[stdlog] 2023-05-25 09:22:10,224 avocado.utils.process DEBUG| [stdout] package ansible is not installed
[stdlog] 2023-05-25 09:22:10,226 avocado.utils.process INFO | Command 'rpm -q ansible' finished with 1 after 0.027414360s
[stdlog] 2023-05-25 09:22:10,227 avocado.utils.process INFO | Running '/bin/dnf -y install ansible'
[stdlog] 2023-05-25 09:22:12,537 avocado.utils.process DEBUG| [stdout] Updating Subscription Management repositories.
(...)
[stdlog] 2023-05-25 09:22:41,984 avocado.utils.process DEBUG| [stdout] Complete!
[stdlog] 2023-05-25 09:22:42,346 avocado.utils.process INFO | Command '/bin/dnf -y install ansible' finished with 0 after 32.104707449s

```